### PR TITLE
chore(completion) - generate shell script in a much more automated way

### PR DESCRIPTION
**Pull request by catull**
All the options are taken straight from the commands, aliases as well.

The `ng generate` command is not fully supported yet.
It is being considered for the main `ng` options, however since it does not have `avaliableOptions` but rather an aliasMap, it requires more work.

### CODE REVIEW

**Comment by catull** [comment]
PR is large, it introduces command options.
Would like to get some feed-back.

**Comment by catull** [comment]
Why is the file packages/angular-cli/utilities/completion.sh conflicting ?
The PR removes it, generates its content on-the-fly.

File: packages/angular-cli/commands/completion.ts [hunk 1]
```diff
@@ -1,17 +1,176 @@
...
+        let opts = extractOptions (command.availableOptions);
+        caseBlock = caseBlock + '    ' + com.sort().join('|') + ') opts="' + opts + '" ;;\n';
+      }
+    });
+
+    caseBlock = 'ng|help) opts="' + optsNg.sort().join(' ') + '" ;;\n' +
+                caseBlock +
+                '    *) opts="" ;;';
+
+    console.log(`###-begin-ng-completion###
```
Comment (hansl) [inline packages/angular-cli/commands/completion.ts]: Could you use `stripIndent` from `common-tags` and indent the whole string? Look at other places where we use it for an example.

File: packages/angular-cli/commands/completion.ts [hunk 2]
```diff
@@ -1,17 +1,176 @@
...
+#
+`);
+
+    if (commandOptions.all && !commandOptions.bash) {
+      console.log('if test ".$(type -t complete 2>/dev/null || true)" = ".builtin"; then');
+    }
+
+    if (commandOptions.all || commandOptions.bash) {
+      console.log(`_ng_completion() {
+  local cword pword opts
```
Comment (hansl) [inline packages/angular-cli/commands/completion.ts]: Same here; use `stripIndent` so we keep the indentation in the code (I don't care much about the generated files).

File: README.md [hunk 3]
```diff
@@ -289,19 +289,19 @@ To turn on auto completion use the following commands:
 
 For bash:
 ```bash
-ng completion 1>> ~/.bashrc 2>>&1
+ng completion --bash 1>> ~/.bashrc
```
Comment (catull) [inline README.md]: Okay, changed it in completion.ts, too.

File: packages/angular-cli/commands/completion.ts (hunk 1 above)
Comment (catull) [inline packages/angular-cli/commands/completion.ts]: Done.

File: packages/angular-cli/commands/completion.ts (hunk 2 above)
Comment (catull) [inline packages/angular-cli/commands/completion.ts]: Done, too.

**Comment by hansl** [comment]
LGTM. Awesome!

**Comment by hansl** [comment]
Waiting for the e2e tests to complete then I'll merge this in time for the beta

**Comment by Brocco** [comment]
I have a request (which can be a separate PR) to extract the availableOptions into a separate file so it can be used by other things than just completion. And with that add the ability to parse out the availableOptions from blueprints as well.

**Comment by catull** [comment]
@Brocco @hansl 
Does this mean this PR is put on hold until the factoring out has happened and the PR is updated accordingly ?

**Comment by hansl** [comment]
No I'm fine with this PR as it is today. Mike does that impede your work?

**Comment by catull** [comment]
I can open a new issue to introduce a separation of availableOptions / blueprints into, say _.options.ts.

**Comment by catull** [comment]
@Brocco New PR https://github.com/angular/angular-cli/pull/4267 has the separated options / blueprints.

**Pull request closed without merging**